hw-model/src/model_emulated.rs

```diff
@@ -262,4 +266,13 @@ impl crate::HwModel for ModelEmulated {
         self.cpu.warm_reset();
         self.step();
     }
+
+    fn cold_reset(&mut self) {
+        // TODO: Only saving ROM from init params right now
+        *self = ModelEmulated::new_unbooted(InitParams {
+            rom: &self.cold_reset_rom,
```

Comment:
I need a better solution for this. Right now this just saves the ROM. Should I just choose other params from InitParams that make sense to save? These are the options I considered and the problems with each.

**1. Store InitParams when we call new_unbooted. This is challenging because**

- InitParams contains several references, we can’t/don’t expect the objects they reference to outlive the model. We would need to store a copy of them somewhere that lives longer and reference that.
- There are several Box<dyn<…>> fields in InitParams that look messy to copy, but I think this is doable without pulling a new crate (dyn-clone) in?
- Some InitParams aren’t logical to carry over through a cold reset (ICCM, DCCM, trng nibbles/responses)

**2. Implement cold reset within the model implementation. This would actually reset reg values within the same instance**

- This looks like a large task and doesn’t seem like the right approach.

**3. Store the fresh copy of the CPU instance in model_emulated as a reset state in new_unbooted.**

- I tried to get CPU to derive clone, but eventually ended up in low-level components that were types from crates that did not or could not derive clone.

**4. Store the fields from InitParams we care about in the model_emulated struct and use them to create a new model in cold_reset.**

- Adds a maintenance cost if more fields are added to InitParams since we manually cherry pick fields to copy
- Adding an InitParam struct within ModelEmulated forces a lifetime annotation (due to the references in InitParams) this propagates all the way up to hw-model and makes a mess. Instead we’d probably want to individually duplicate the fields we want. The references in InitParams are not really useful anyway since we have to store a full copy of the data somewhere.

Reply:
Direction for now is to not support this in the SW emulator and push that decision making to the test. There can be a either be switch for different behavior in the sw-emulator instance (to create a new model) or the test can just be allowed to run on verilator and FPGA only